### PR TITLE
TracerConverter implementation to automatically wrap a tracer with th…

### PR DESCRIPTION
**Pull request by objectiser**
…e APIExtensionsTracer and add any available TracerObserver implementations resolved through service loaded TracerObserverResolvers

### CODE REVIEW

**Comment by objectiser** [comment]
@sjoerdtalsma Would you mind reviewing this, as it is using/extending tracer resolver?

For context, basically the `APIExtensionsTracer` is a generic tracer wrapper that can be used to add callbacks to - initially just an observer API. So this PR is adding a converter, to automatically wrap the resolved tracer with this `APIExtensionsTracer`, as well as use the service loader mechanism to resolve observers from the classpath.

**Comment by sjoerdtalsma** [comment]
I hope I can make some time before the weekend. I have some social event I must attend though.

**Comment by objectiser** [comment]
@sjoerdtalsma Thanks - no rush, next week is fine.

**Comment by pavolloffay** [comment]
Wouldn't it be better to call it tracer decorator? Is it converting the tracer to something else? Or does it add a behavior?

**Comment by objectiser** [comment]
I'll leave that as a question for @sjoerdtalsma - as that would require a change in tracerresolver.

**Comment by sjoerdtalsma** [comment]
About the question, @pavolloffay 
>Wouldn't it be better to call it tracer decorator? Is it converting the tracer to something else? Or does it add a behavior?

First of all; I'm not a native English speaker.
I think a decorator is narrower than a converter. A converter, well, 'converts' one tracer into another and may decorate, but may also fundamentally alter behavior in other ways.  _edit: For example, the `TracerResolver` even supports converters that act as 'filter': when a `Tracer` is converted to `null` it is skipped (even if it has highest priority) and the next one will be resolved._

However, if this particular implementation clearly is a decorator and wish to name it as such, I see no problem doing so, while implementing (and registering) it as a `TracerConverter` SPI.
I don't really see why that would require a change in the tracer resolver?

File: opentracing-api-extensions-tracerconverter/src/main/java/io/opentracing/contrib/api/tracer/converter/APIExtensionsTracerConverter.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            APIExtensionsTracer extTracer = new APIExtensionsTracer(tracer);
+            for (TracerObserver observer : observers) {
+                extTracer.addTracerObserver(observer);
+            }
+            tracer = extTracer;
+        }
+        return tracer;
+    }
+
+    List<TracerObserver> getObservers() {
```
Comment (sjoerdtalsma) [inline opentracing-api-extensions-tracerconverter/src/main/java/io/opentracing/contrib/api/tracer/converter/APIExtensionsTracerConverter.java]: Can be `private static`

File: opentracing-api-extensions-tracerconverter/src/main/java/io/opentracing/contrib/api/tracer/converter/APIExtensionsTracerConverter.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import io.opentracing.contrib.api.tracer.APIExtensionsTracer;
+import io.opentracing.contrib.tracerresolver.TracerConverter;
+
+/**
+ * This implementation of the {@link TracerConverter} interface is used to create an
+ * {@link APIExtensionsTracer} wrapper, if one or more {@link TracerObserver} implementation can
+ * be resolved using service loadable {@link TracerObserverResolver}s.
+ *
+ */
+public class APIExtensionsTracerConverter implements TracerConverter {
```
Comment (sjoerdtalsma) [inline opentracing-api-extensions-tracerconverter/src/main/java/io/opentracing/contrib/api/tracer/converter/APIExtensionsTracerConverter.java]: Maybe rename to `APIExtensionsTracerDecorator` and still implement `TracerConverter` (inspired by @pavolloffay's comment) ?

**Review by sjoerdtalsma — COMMENTED** [review]
I have see no problem with this implementation.
Maybe think about naming; converter vs. decorator (also in modules / packages) before merging this.
Also is the indirection through `TracerObserverResolver` warranted? 
You could also just do a SPI lookup for `TracerObserver` itself. Maybe explain the arguments for this decision in the Javadoc a little clearer?

**Review by sjoerdtalsma — APPROVED** [review]

**Comment by objectiser** [comment]
@sjoerdtalsma @pavolloffay Thanks - looks like I had forgotten this one for a long time :)

**Pull request merged**